### PR TITLE
Deltas for 'wrangler pages publish'

### PR DESCRIPTION
**Pull request by GregBrimble**
~Waiting on an internal thing to be released, but once that's out, this PR will be ready to go.~ It's out!

feat: Upload the delta for `wrangler pages publish`

We now keep track of the files that make up each deployment and intelligently only upload the files that we haven't seen. This means that similar subsequent deployments should only need to upload a minority of files and this will hopefully make uploads even faster.

The only bit that has changed is this: https://github.com/cloudflare/wrangler2/pull/1210/files#diff-671053f3ecffd0e36229ae8b3c6090a2becd125b2c3247a41de85836dca29884R1063-R1078

The rest is just pulling out the upload logic for future re-use.

### CODE REVIEW

File: packages/wrangler/src/pages.tsx [hunk 1]
```diff
@@ -821,6 +821,270 @@ interface CreateDeploymentArgs {
...
+        if (filestat.isSymbolicLink()) {
+          return;
+        }
+
+        if (filestat.isDirectory()) {
+          fileMap = await walk(filepath, fileMap, depth + 1);
+        } else {
+          let name;
+          if (depth) {
+            name = filepath.split(sep).slice(1).join("/");
```
Comment (jrf0110) [inline packages/wrangler/src/pages.tsx:874]: I know this is just moved code, but we should be joining by `sep` here

File: packages/wrangler/src/pages.tsx [hunk 2]
```diff
@@ -821,6 +821,270 @@ interface CreateDeploymentArgs {
...
+
+  unmount();
+
+  const uploadMs = Date.now() - start;
+
+  logger.log(
+    `✨ Success! Uploaded ${fileMap.size} files ${formatTime(uploadMs)}\n`
+  );
+
+  try {
```
Comment (sidharthachatterjee) [inline packages/wrangler/src/pages.tsx:1063]: Does this need a retry? For instance if the JWT expired by then? (Like we have for the calls to `/upload`

File: packages/wrangler/src/pages.tsx (hunk 1 above)
Comment (GregBrimble) [reply packages/wrangler/src/pages.tsx:874]: No, we don't, because we need to replace Windows separators with UNIX ones.

**Review by sidharthachatterjee — APPROVED** [review]
Looking good! Let's ensure the new API is up before we get this in. Thank you for the refactoring as well. That'll be helpful for later.

File: packages/wrangler/src/pages.tsx (hunk 1 above)
Comment (sidharthachatterjee) [reply packages/wrangler/src/pages.tsx:874]: Yeah, we want to always join with `/` because we are constructing url paths for lookup here effectively.

**Comment by jrf0110** [comment]
Fantastic!!

We may consider tweaking `MAX_BUCKET_SIZE`. In my testing, there was a period of around 30-50s with no progress being reported

File: packages/wrangler/src/pages.tsx (hunk 1 above)
Comment (jrf0110) [reply packages/wrangler/src/pages.tsx:874]: # 🤦 

Duh. Right! Carry on :)

File: packages/wrangler/src/pages.tsx (hunk 2 above)
Comment (GregBrimble) [reply packages/wrangler/src/pages.tsx]: Added a retry 👍

**Comment by GregBrimble** [comment]
> We may consider tweaking MAX_BUCKET_SIZE. In my testing, there was a period of around 30-50s with no progress being reported

We can definitely do this if it's a commonly seen problem (or at the very least, we can provide more obvious feedback that an upload is actually happening). I'll keep an eye on issues. If you routinely see this, file one yourself. Otherwise, I'll wait to see if anyone else creates one.

**Pull request merged**